src/statsd-parsers/statsd-parsers.h

```diff
@@ -1,23 +1,33 @@
+#ifndef STATSD_PARSERS_
+#define STATSD_PARSERS_
+
 #include <chan/chan.h>
 
 #include "../config-reader/config-reader.h"
 
-#ifndef STATSD_PARSERS_
-#define STATSD_PARSERS_
+typedef struct metrics metrics;
 
 const int PARSER_TRIVIAL;
 const int PARSER_RAGEL;
 
+typedef struct tag {
+    char* key;
+    char* value;
+} tag;
+
+typedef struct tag_collection {
+    tag** values;
+    long int length;
+} tag_collection;
+
 typedef struct statsd_datagram
 {
-    char *data_namespace;
-    char *type;
-    char *modifier;
-    char *tags;
-    double value;
-    char *metric;
-    char *instance;
-    char *sampling;
+    char* metric;
+    char* type;
+    tag_collection* tags;
```

Comment:
@natoscott does it make sense to actually parse labels (they are in form of "label=value,label2=value2")? These structures can be complex in memory and it might not make huge sense to parse then since they can only be used as input into hash (metric name + labels string). The agent could actually only verify if the order is correct (alphabetical order) and if so then we would simply keep them in memory as string. Only when passing them into PCP agent would parse them.

OR

Alternatively, since PCP requires them to be in JSON, the agent would actually parse them early and store as JSON in memory. Hashing could be done over JSON string easily, there is not much data added.

One or the other, the tag could be simple string not a pointer to array. This could simplify storage and handling. Not sure.

Reply:
@lzap We talked about saving tagged values as "children" of "naked" metric values. These children would also have reference back to parent, so it would be easy to get all records of given metric name tagged or not. I don't see how that would be possible with JSON without the need to explicitly parse tag keys as well.

Reply:
Sure, it's flexible storage however the question is - will user be able to list metrics by labels? This is just an agent which forwards everything into PCP, I don't think that the effort and complexity of the relevant code is worth it.

However, this is fine if you want to keep this as is.